runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -591,13 +602,40 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
         ctx.close();
     }
 
+    /**
+     * Sends a ping to the server so that the pong response will keep
+     * the channel active in order to avoid a ReadTimeout exception that will
+     * close the channel.
+     */
+    private void keepAlive() {
+        if (!channel.isOpen()) {
+            log.warn("keepAlive: channel not open, skipping ping. ");
+            return;
+        }
+        channel.writeAndFlush(new CorfuMsg(CorfuMsgType.PING), channel.voidPromise());
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (keepAliveTask != null) {
+            log.debug("channelInactive: Canceling keep-alive task");
+            keepAliveTask.cancel(true);
+            keepAliveTask = null;
+        }
+
+        // Should the handshake handler also do this?
```

Comment:
This sends a ping even if the channel is active ? Can we optimize this (if not now then in the future). This can generate extra load (this is all in theory at the moment, I have nothing to substantiate this) on server when there are lots of clients.

Reply:
That's correct. We can change this so that it fires when there has been no writes for sometime

Reply:
Done. Changed it to only trigger when there is no write activity (i.e. won't send keep-alive messages where there is write activity on the channel .